File: SAAF/java_template/src/main/java/lambda/ExtractTransform.java
package lambda;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.GetObjectRequest;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.amazonaws.services.s3.model.S3Object;
import saaf.Inspector;
import saaf.Response;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.io.StringWriter;
import java.math.BigDecimal;
import java.nio.charset.StandardCharsets;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.*;

public class ExtractTransform implements RequestHandler<Request, HashMap<String, Object>> {
    @Override
    public HashMap<String, Object> handleRequest(Request request, Context context) {
        //Collect initial data.
        Inspector inspector = new Inspector();
        inspector.inspectAll();

//        LambdaLogger logger = context.getLogger();

        String bucketname = request.getBucketname();
        String filename = request.getFilename();
//        logger.log("Received bucketname:" + bucketname + " Received filename:" + filename);

        AmazonS3 s3Client = AmazonS3ClientBuilder.standard().build();

        S3Object s3Object = s3Client.getObject(new GetObjectRequest(bucketname, filename));

        InputStream objectData = s3Object.getObjectContent();

        HashSet<Sale> sales = new HashSet<>();

        // Read sales data from CSV
        Scanner scanner = new Scanner(objectData);
        while (scanner.hasNext()) {
            String line = scanner.nextLine();
            // Skips column names
            if (line.startsWith("Region,Country,")) {
                continue;
            }
            Sale newSale = getSale(line);

            sales.add(newSale);
        }
        scanner.close();

        StringWriter sw = new StringWriter();
        // Add column names
        sw.append("Region,Country,Item Type,Sales Channel,Order Priority,Order Date,Order ID,Ship Date,Units Sold,Unit Price,Unit Cost,Total Revenue,Total Cost,Total Profit,Order Processing Time,Gross Margin\n");
        // Write sales data to new CSV
        for (Sale s : sales) {
            sw.append(s.toString()).append("\n");
        }

        byte[] bytes = sw.toString().getBytes(StandardCharsets.UTF_8);
        InputStream is = new ByteArrayInputStream(bytes);
        ObjectMetadata meta = new ObjectMetadata();
        meta.setContentLength(bytes.length);
        meta.setContentType("text/plain");

        AmazonS3 s3ClientOutput = AmazonS3ClientBuilder.standard().build();
        // Creates a new file Name;
        int dotIndex = filename.lastIndexOf(".");
        String newFileName = filename.substring(0, dotIndex) + "-et" + filename.substring(dotIndex);

        s3ClientOutput.putObject(bucketname, newFileName, is, meta);

        Response response = new Response();
        response.setValue("Bucket:" + bucketname + " filename:" + newFileName + " size:" + bytes.length);

        inspector.consumeResponse(response);

        //Collect final information such as total runtime and cpu deltas.
        inspector.inspectAllDeltas();
        return inspector.finish();
    }

    private static Sale getSale(String line) {
        String[] data = line.split(",");
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("M/d/yyyy");

        return new Sale(
                data[0],
                data[1],
                data[2],
                data[3],
                data[4],
                LocalDate.parse(data[5], formatter),
                Integer.parseInt(data[6]),
                LocalDate.parse(data[7], formatter),
                Integer.parseInt(data[8]),
                new BigDecimal(data[9]),
                new BigDecimal(data[10]),
                new BigDecimal(data[11]),
                new BigDecimal(data[12]),
                new BigDecimal(data[13]));
    }
}

File: SAAF/python_template/src/extract_transform.py
import boto3
import json
import io
import os
from decimal import Decimal
from datetime import datetime

from Inspector import Inspector   # your provided Inspector code
from sale import Sale             # Sale model above


DATE_FORMAT = "%m/%d/%Y"


def load_database_config():
    """
    Java version reads db.properties.
    In AWS Lambda, environment variables are the correct equivalent.
    """
    url = os.getenv("DB_URL")
    if not url:
        raise RuntimeError("DB_URL environment variable not set")
    return url


def parse_sale(line):
    data = line.split(",")

    return Sale(
        region=data[0],
        country=data[1],
        item_type=data[2],
        sales_channel=data[3],
        order_priority=data[4],
        order_date=datetime.strptime(data[5], DATE_FORMAT).date(),
        order_id=int(data[6]),
        ship_date=datetime.strptime(data[7], DATE_FORMAT).date(),
        units_sold=int(data[8]),
        unit_price=Decimal(data[9]),
        unit_cost=Decimal(data[10]),
        total_revenue=Decimal(data[11]),
        total_cost=Decimal(data[12]),
        total_profit=Decimal(data[13])
    )


def lambda_handler(event, context):
    inspector = Inspector()
    inspector.inspectAll()

    bucketname = event["bucketname"]
    filename = event["filename"]

    # Load DB config
    url = load_database_config()

    s3 = boto3.client("s3")
    response = s3.get_object(Bucket=bucketname, Key=filename)

    sales = set()

    body = io.TextIOWrapper(response["Body"], encoding="utf-8")
    for line in body:
        line = line.strip()
        if line.startswith("Region,Country,"):
            continue
        sales.add(parse_sale(line))

    # Write transformed CSV
    output = io.StringIO()
    output.write(
        "Region,Country,Item Type,Sales Channel,Order Priority,Order Date,"
        "Order ID,Ship Date,Units Sold,Unit Price,Unit Cost,"
        "Total Revenue,Total Cost,Total Profit,Order Processing Time,Gross Margin\n"
    )

    for sale in sales:
        output.write(str(sale) + "\n")

    csv_bytes = output.getvalue().encode("utf-8")

    dot = filename.rfind(".")
    new_filename = f"{filename[:dot]}-et{filename[dot:]}"

    s3.put_object(
        Bucket=bucketname,
        Key=new_filename,
        Body=csv_bytes,
        ContentType="text/plain"
    )

    response_payload = {
        "bucketname": bucketname,
        "filename": new_filename,
        "size": len(csv_bytes)
    }

    inspector.addAttribute("bucketname", bucketname)
    inspector.addAttribute("filename", new_filename)
    inspector.addAttribute("dbUrl", url)
    inspector.addAttribute("loadError", Decimal("0"))
    inspector.addAttribute("response", json.dumps(response_payload))

    inspector.inspectAllDeltas()
    return inspector.finish()